src/Jaeger.Core/Reporters/RemoteReporter.cs

```diff
@@ -62,52 +60,31 @@ public void Report(Span span)
             }
         }
 
-        private void ConsumeQueue()
+        public async Task CloseAsync(CancellationToken cancellationToken)
         {
-            while (!_commandQueue.IsCompleted)
-            {
-                try
-                {
-                    // This blocks until a span is available.
-                    ICommand command = _commandQueue.Take();
-
-                    try
-                    {
-                        command.Execute();
-                    }
-                    catch (SenderException ex)
-                    {
-                        _metrics.ReporterFailure.Inc(ex.DroppedSpanCount);
-                    }
-                }
-                catch (Exception ex)
-                {
-                    _logger.LogError(ex, "QueueProcessor error");
-                    // Do nothing, and try again on next span.
-                }
-            }
-        }
+            // Note: Java creates a CloseCommand but we have CompleteAdding() in C# so we don't need the command.
+            // (This also stops the FlushLoop)
+            _commandQueue.CompleteAdding();
 
-        public void Dispose()
-        {
             try
             {
-                // Note: Java creates a CloseCommand but we have CompleteAdding() in C# so we don't need the command.
-                _commandQueue.CompleteAdding();
+                // Give processor some time to process any queued commands.
 
-                // Give sender some time to process any queued spans.
-                _queueProcessor.Wait(10000);
+                var cts = CancellationTokenSource.CreateLinkedTokenSource(
+                    cancellationToken,
+                    new CancellationTokenSource(10000).Token);
+
+                _queueProcessorTask.Wait(cts.Token);
```

Comment:
`_flushTask` is never awaited. Maybe we should await it here too after the `_queueProcessorTask`? The `_commandQueue.CompleteAdding()` above should lead the `FlushLoop` to be closed after max. `_flushInterval`. Also I don't like to have a sync `Wait` in an async method. I would propose the following:

```csharp
var cts = CancellationTokenSource.CreateLinkedTokenSource(
    cancellationToken,
    new CancellationTokenSource(TimeSpan.FromSeconds(10)).Token);
var cancellationTask = Task.Delay(Timeout.Infinite, cts.Token);

await Task.WhenAny(_queueProcessorTask, cancellationTask);
await Task.WhenAny(_flushTask, cancellationTask);
```

The `Task.Delay` variant to create a cancellationTask can also be found in the [MSDN Blog](https://blogs.msdn.microsoft.com/cesardelatorre/2017/11/18/implementing-background-tasks-in-microservices-with-ihostedservice-and-the-backgroundservice-class-net-core-2-x/). Another option would be the [CancellationTokenTaskSource](https://github.com/StephenCleary/AsyncEx/blob/master/src/Nito.AsyncEx.Tasks/CancellationTokenTaskSource.cs) from Stephen Cleary. A third option would be to not use `CreateLinkedTokenSource` but only `Task.Delay(10000, cancellationToken)` and check if the result of `await Task.WhenAny` is `cancellationTask`, but in that case no `OperationCanceledException` would be thrown on timeout, only on `cancellationToken`. That's why I prefer the first two options.